### PR TITLE
feat: enable feature store online serving

### PR DESCRIPTION
**Pull request by morgandu**
- [x] Added `_instantiate_featurestore_online_client`, and `read` method in `EntityType` class to enable feature store online serving
- [x] Enabled construct Pandas DataFrame for online serving result
- [x] Added unit tests
- [x] Added integration tests

### CODE REVIEW

**Review by sasha-gitg — CHANGES_REQUESTED** [review]

File: google/cloud/aiplatform/featurestore/entity_type.py [hunk 1]
```diff
@@ -1229,54 +1223,63 @@ def read(
...
-        """Loads read_feature_values_response to dataframe
 
-        Args:
-            response (Union[featurestore_online_service.ReadFeatureValuesResponse,
-                            List[featurestore_online_service.ReadFeatureValuesResponse]):
-                Required. featurestore_online_service.ReadFeatureValuesResponse from `read_feature_values`
-                or Iterable in list from `streaming_read_feature_values`.
-        Returns:
-            pd.DataFrame - response in DataFrame
+def _load_read_feature_values_response_to_dataframe(
```
Comment (sasha-gitg) [inline google/cloud/aiplatform/featurestore/entity_type.py:1229]: What was the reasoning for not keeping this a staticmethod? It seems like it's coupled to the class.

**Review by sasha-gitg — APPROVED** [review]
LGTM. Please get approval from Diem/FS team.

File: tests/unit/aiplatform/test_featurestores.py [hunk 2]
```diff
@@ -291,6 +300,66 @@ def import_feature_values_mock():
...
+                        id=_TEST_FEATURE_ID
+                    ),
+                ]
+            ),
+            entity_view=gca_featurestore_online_service.ReadFeatureValuesResponse.EntityView(
+                entity_id=_TEST_READ_ENTITY_ID,
+                data=[
+                    gca_featurestore_online_service.ReadFeatureValuesResponse.EntityView.Data(
+                        value=gca_featurestore_online_service.FeatureValue(
+                            int64_value=0
```
Comment (diemtvu) [inline tests/unit/aiplatform/test_featurestores.py:322]: can we use non trivial value for test?

File: google/cloud/aiplatform/featurestore/entity_type.py [hunk 3]
```diff
@@ -1138,3 +1145,141 @@ def ingest_from_gcs(
...
+        return initializer.global_config.create_client(
+            client_class=utils.FeaturestoreOnlineServingClientWithOverride,
+            credentials=credentials,
+            location_override=location,
+        )
+
+    def read(
+        self,
+        entity_ids: Union[str, List[str]],
+        feature_ids: Optional[Union[str, List[str]]] = None,
```
Comment (diemtvu) [inline google/cloud/aiplatform/featurestore/entity_type.py:1174]: nit. Is this better to make the default to "*" and drop the Optional?

Comment (morgandu) [reply google/cloud/aiplatform/featurestore/entity_type.py:1174]: done

File: google/cloud/aiplatform/featurestore/entity_type.py (hunk 1 above)
Comment (morgandu) [reply google/cloud/aiplatform/featurestore/entity_type.py:1229]: changed to staticmethod

File: tests/unit/aiplatform/test_featurestores.py (hunk 2 above)
Comment (morgandu) [reply tests/unit/aiplatform/test_featurestores.py]: done

File: google/cloud/aiplatform/featurestore/entity_type.py [hunk 4]
```diff
@@ -157,7 +164,7 @@ def update(
         self,
         description: Optional[str] = None,
         labels: Optional[Dict[str, str]] = None,
-        request_metadata: Optional[Sequence[Tuple[str, str]]] = (),
+        request_metadata: Sequence[Tuple[str, str]] = (),
```
Comment (diemtvu) [inline google/cloud/aiplatform/featurestore/entity_type.py:167]: For my education, why is this required? What is the request_metadata btw?

File: google/cloud/aiplatform/featurestore/entity_type.py [hunk 5]
```diff
@@ -1138,3 +1145,144 @@ def ingest_from_gcs(
...
+            for feature_descriptor in response_header.feature_descriptors
+        ]
+        data = []
+
+        for entity_view in response_entity_view:
+            entity_id = entity_view.entity_id
+            entity_data = [entity_id]
+
+            for entity_view_data in entity_view.data:
+                if entity_view_data._pb.HasField("value"):
```
Comment (diemtvu) [inline google/cloud/aiplatform/featurestore/entity_type.py:1279]: `data` can be empty (i.e feature missing value). Should we append(None) in that case to make sure entity_data alignment?

File: tests/system/aiplatform/test_featurestore.py [hunk 6]
```diff
@@ -247,3 +249,19 @@ def test_search_features(self, shared_state):
...
+
+    def test_online_reads(self, shared_state):
+        assert shared_state["user_entity_type"]
+        assert shared_state["movie_entity_type"]
+
+        user_entity_type = shared_state["user_entity_type"]
+        movie_entity_type = shared_state["movie_entity_type"]
+
+        user_entity_views = user_entity_type.read(entity_ids="alice")
+        assert type(user_entity_views) == pd.DataFrame
```
Comment (diemtvu) [inline tests/system/aiplatform/test_featurestore.py:261]: Can we test the actual values?

File: tests/unit/aiplatform/test_featurestores.py [hunk 7]
```diff
@@ -1054,6 +1123,50 @@ def test_ingest_from_gcs_with_invalid_gcs_source_type(self):
                 gcs_source_type=_TEST_GCS_SOURCE_TYPE_INVALID,
             )
 
+    @pytest.mark.usefixtures("get_entity_type_mock")
+    def test_read_single_entity(self, read_feature_values_mock):
```
Comment (diemtvu) [inline tests/unit/aiplatform/test_featurestores.py:1127]: Can we have a test which the response is parsed, i.e have multiple columns/features, some of them are empty?

File: google/cloud/aiplatform/featurestore/entity_type.py (hunk 4 above)
Comment (morgandu) [reply google/cloud/aiplatform/featurestore/entity_type.py:167]: This is where the `request_metadata` will be used:

https://github.com/googleapis/python-aiplatform/blob/main/google/cloud/aiplatform_v1/services/featurestore_service/client.py#L911

When we used `request_metadata: Optional[Sequence[Tuple[str, str]]] = None`, it will throw an error, 

https://github.com/googleapis/python-aiplatform/blob/main/google/cloud/aiplatform_v1/services/featurestore_service/client.py#L993

so we use the same default value '()' as in the GAPIC layer.

File: tests/unit/aiplatform/test_featurestores.py (hunk 7 above)
Comment (morgandu) [reply tests/unit/aiplatform/test_featurestores.py:1127]: added test case for all value type, with missing data value scenario

File: tests/system/aiplatform/test_featurestore.py (hunk 6 above)
Comment (morgandu) [reply tests/system/aiplatform/test_featurestore.py:261]: there is a check actual value a couple lines below, also added another unit test for checking data frame construction, where it has more cases

File: google/cloud/aiplatform/featurestore/entity_type.py (hunk 5 above)
Comment (morgandu) [reply google/cloud/aiplatform/featurestore/entity_type.py:1279]: fixed this!

File: google/cloud/aiplatform/featurestore/entity_type.py [hunk 8]
```diff
@@ -1138,3 +1146,162 @@ def ingest_from_gcs(
...
+                f"Numpy and/or Pandas are not installed. Please install numpy and/or pandas to use "
+                f"{EntityType._construct_dataframe.__name__}"
+            )
+
+        data = []
+        feature_ids = list(feature_metadata.keys())
+        for entity_view in entity_views:
+            entity_data = {"entity_id": entity_view.entity_id}
+            for feature_id, entity_view_data in zip(feature_ids, entity_view.data):
+                if entity_view_data._pb.HasField("value"):
```
[the remark this answers is not in the document]
Comment (morgandu) [reply google/cloud/aiplatform/featurestore/entity_type.py:1284]: Added enforced None for entity_view_data proto withou the "value".

FYI, for data contains None, when the feature value type is "int64", pd will convert the dtype to np.dtype("float64"), with NaN instead of None. For other value types, dtype will be np.dtype("object") with None.

File: tests/unit/aiplatform/test_featurestores.py [hunk 9]
```diff
@@ -1054,6 +1200,193 @@ def test_ingest_from_gcs_with_invalid_gcs_source_type(self):
...
+        )
+
+        expected_data = []
+        for entity_id, entity_feature_values in zip(entity_ids, feature_values):
+            entity_data = {}
+            entity_data["entity_id"] = entity_id
+            for feature_id, feature_value in zip(feature_ids, entity_feature_values):
+                if feature_value is not None:
+                    entity_data[feature_id] = feature_value
+            expected_data.append(entity_data)
```
[the remark this answers is not in the document]
Comment (morgandu) [reply tests/unit/aiplatform/test_featurestores.py]: Sure! Updated tests with explicit df to avoid any potential programmatically bugs.

File: tests/unit/aiplatform/test_featurestores.py [hunk 10]
```diff
@@ -1054,6 +1218,200 @@ def test_ingest_from_gcs_with_invalid_gcs_source_type(self):
...
+                        [True, False],
+                        1.2,
+                        [1.2, 3.4],
+                        1,
+                        [1, 2],
+                        "test",
+                        ["test1", "test2"],
+                        b"1",
+                    ],
+                    [None, None, None, None, None, None, None, None, None],
```
Comment (diemtvu) [inline tests/unit/aiplatform/test_featurestores.py:1342]: nit. Can you change/add a row that have both valid values and None.

**Review by diemtvu — APPROVED** [review]

File: tests/unit/aiplatform/test_featurestores.py (hunk 10 above)
Comment (morgandu) [reply tests/unit/aiplatform/test_featurestores.py:1342]: updated

File: google/cloud/aiplatform/featurestore/entity_type.py [hunk 11]
```diff
@@ -1138,3 +1145,144 @@ def ingest_from_gcs(
...
+        Returns:
+            pd.DataFrame - entities feature values in DataFrame
+        )
+        """
+
+        try:
+            import pandas as pd
+        except ImportError:
+            raise ImportError(
+                f"Pandas are not installed. Please install pandas to use "
```
Comment (kweinmeister) [inline google/cloud/aiplatform/featurestore/entity_type.py:1270]: Pandas library *is* not installed?

**Review by kweinmeister — APPROVED** [review]

**Comment by ivanmkc** [comment]
Missing PR description?
